File: erpnext/erpnext/leadership_archival/report/leadership_archival__master_report/leadership_archival__master_report.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": "Name", "fieldname": "recipientName", "fieldtype": "Data", "width": 150},
        {"label": "CID", "fieldname": "cid", "fieldtype": "Data", "width": 150},
        {"label": "DoB", "fieldname": "dob", "fieldtype": "phone", "width": 150},
        {"label": "Contact No", "fieldname": "contact_no", "fieldtype": "phone", "width": 150},
        {"label": "Title", "fieldname": "title", "fieldtype": "Data", "width": 150},
        {"label": "Location", "fieldname": "location", "fieldtype": "Data", "width": 150},
        {"label": "Event Name", "fieldname": "event_name", "fieldtype": "Data", "width": 150},
        {"label": "Confer By","options": "Conferred By", "fieldname": "conferred_by", "fieldtype": "Link", "width": 150},
        {"label": "Citation", "fieldname": "citation", "fieldtype": "Data", "width": 150},
        {"label": "Award or Appointment Date", "fieldname": "issued_date", "fieldtype": "Data", "width": 150},
        {"label": "PostThumous","fieldname": "postThumous", "fieldtype": "check", "width": 150},
        {"label": "Link to Profile","options": "Key Person Registry", "fieldname": "profile", "fieldtype": "Link", "width": 150},
        {"label": "Kasho","options": "Kasho", "fieldname": "kasho", "fieldtype": "Link", "width": 150},
    ]
    
def get_data(filters):
    conditions = []

    if filters.get("cid"):
        conditions.append("aw.cid = %(cid)s")

    if filters.get("title_medal"):
        conditions.append("aw.title = %(title_medal)s")
    # ✅ FORCE Kasho Type (no frontend dependency)
    conditions.append("k.kasho_type IN ('Award', 'Appointment')")
        
    start_date = filters.get("start_date")
    end_date = filters.get("end_date")

    if start_date and end_date:
        if end_date < start_date:
            frappe.throw("End Date must be greater than Start Date")
        conditions.append("k.issue_date BETWEEN %(start_date)s AND %(end_date)s")
    elif start_date:
        conditions.append("k.issue_date >= %(start_date)s")
    elif end_date:
        conditions.append("k.issue_date <= %(end_date)s")

    where_clause = " AND ".join(conditions)

    if where_clause:
        where_clause = "WHERE " + where_clause
    else:
        return []  # 🚨 No filters → no data

    return frappe.db.sql(f"""
        SELECT 
			aw.cid AS cid,
			kpr.registry_name AS recipientName,
			kpr.dob AS dob,
			aw.title AS title,
			aw.location AS location,
			aw.event_name AS event_name,
			aw.conferred_by AS conferred_by,
			aw.citation AS citation,
			k.issue_date AS issued_date,
			aw.citation AS citation,
			k.name AS kasho,
			kpr.name AS profile,
			aw.postThumous AS postThumous
		FROM `tabAward and Appointment` aw
		INNER JOIN `tabKasho` k 
			ON aw.parent = k.name
		LEFT JOIN `tabKey Person Registry` kpr 
			ON kpr.cid = aw.cid
        {where_clause}
    """, filters, as_dict=1)

File: erpnext/erpnext/leadership_archival/report/leadership_archival__master_report/leadership_archival__master_report.js
// Copyright (c) 2026, Frappe Technologies and contributors
// For license information, please see license.txt

frappe.query_reports["Leadership Archival  Master Report"] = {

    "filters": [
        {
            "fieldname": "cid",
            "label": "CID",
            "fieldtype": "Data",
            "placeholder": "Enter CID"
        },
        {
            "fieldname": "title_medal",
            "label": "Title or Medal",
            "fieldtype": "Link",
            "options": "Title",
            "placeholder": "Select title or Medal"
        },
        {
            "fieldname": "start_date",
            "label": "Start Date",
            "fieldtype": "Date",
            "placeholder": "Start Date"
        },
        {
            "fieldname": "end_date",
            "label": "End Date",
            "fieldtype": "Date",
            "placeholder": "End Date"
        }
    ],


    // =====================================================
    // FORMATTER
    // =====================================================

    formatter: function(value, row, column, data, default_formatter) {

        value = default_formatter(
            value,
            row,
            column,
            data
        );

        // Only Profile column
        if (
            column.fieldname === "profile" &&
            data &&
            data.profile
        ) {
            value = `
                <a
                    href="/app/leadership-profile?profile=${encodeURIComponent(data.profile)}"
                    class="leadership-profile-link"
                    data-profile="${frappe.utils.escape_html(data.profile)}"
                >
                    ${frappe.utils.escape_html(data.profile)}
                </a>
            `;
        }
        return value;
    },


    // =====================================================
    // ONLOAD
    // =====================================================

    onload: function(report) {


        // =================================================
        // PROFILE CLICK
        // =================================================

        $(report.page.wrapper).on(
            "click",
            ".leadership-profile-link",
            function(e) {

                e.preventDefault();
                e.stopPropagation();

                const profile =
                    $(this).attr("data-profile");

                console.log("================================");
                console.log("PROFILE CLICKED");
                console.log("Profile:", profile);
                console.log("================================");


                frappe.set_route(
                    "leadership-profile",
                     profile
                );

            }
        ); // <-- THIS WAS MISSING


        // =================================================
        // DATE VALIDATION
        // =================================================

        function validate_dates() {

            const start =
                report.get_filter_value("start_date");

            const end =
                report.get_filter_value("end_date");


            if (
                start &&
                end &&
                end < start
            ) {

                frappe.msgprint({
                    title: __("Invalid Date Range"),
                    message: __("End Date must be greater than Start Date"),
                    indicator: "red"
                });

                report.set_filter_value(
                    "end_date",
                    null
                );

                return false;
            }

            return true;
        }

        // =================================================
        // REFRESH REPORT
        // =================================================

        function refresh_report() {

            if (!validate_dates()) {
                return;
            }
            frappe.show_progress(
                __("Loading Report"),
                50,
                100,
                __("Fetching data...")
            );
            report.refresh();
            frappe.after_ajax(() => {
                frappe.show_progress(
                    __("Loading Report"),
                    100,
                    100,
                    __("Done")
                );
                setTimeout(() => {

                    frappe.hide_progress();

                }, 300);
            });
        }

        // =================================================
        // BIND FILTERS
        // =================================================

        setTimeout(() => {
            const start_filter =
                report.get_filter("start_date");
            const end_filter =
                report.get_filter("end_date");
            if (
                start_filter &&
                end_filter
            ) {

                start_filter.$input.on(
                    "change",
                    refresh_report
                );

                end_filter.$input.on(
                    "change",
                    refresh_report
                );
            }
        }, 500);
        // =================================================
        // INITIAL LOAD
        // =================================================

        refresh_report();
    }
};